### PR TITLE
[FIX] models.py: Fix check_company mechanism

### PR DESCRIPTION
**Pull request by tivisse**
Purpose
=======

If check company is set on a field and if the user has no access
right to the field value (example: address_id on an expense sheet
could be set to a private res.partner, on an onchange method when
setting the employee), then the check_company mechanism will
raise an AccessError when trying the validate the companies on
the different records.

Specification
=============

As we only wish to validate the new record values and not the
access rights, the validation could be done as a superuser to
avoid unecessary errors.

TaskID: 2170006

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by tivisse** [comment]
@rco-odoo @sle-odoo What do you think about this one ? 

Thanks.

**Comment by sle-odoo** [comment]
cc @Feyensv

File: odoo/models.py [hunk 1]
```diff
@@ -3151,7 +3151,7 @@ def _check_company(self, fnames=None):
                     if not (company <= record[name].company_ids):
                         inconsistent_fields.add(name)
                         inconsistent_recs |= record
-                elif not (record[name].company_id <= company):
+                elif not (record.sudo()[name].company_id <= company):
```
Comment (rco-odoo) [inline odoo/models.py]: The cases above (`record[name].company_ids`) are not protected.
I suggest to factor out `record[name]` like
```python
corecord = record.sudo()[name]
```
and use `corecord` in the 4 places.

File: odoo/models.py [hunk 2]
```diff
@@ -3168,7 +3168,7 @@ def _check_company(self, fnames=None):
...
```
Comment (rco-odoo) [inline odoo/models.py]: Same remark as above.

File: addons/hr_expense/tests/test_expenses.py [hunk 3]
```diff
@@ -346,3 +346,41 @@ def test_expense_refuse(self):
...
+            'product_id': product.id,
+            'unit_amount': 700.00,
+        })
+
+        # Even if the employee's home address is private, the
+        # report creation shouldn't crash.
+        with Form(self.env['hr.expense.sheet'].with_user(user)) as sheet_form:
+            sheet_form.expense_line_ids.add(expense)
+            sheet_form.employee_id = employee
+            sheet_form.name = 'My Sheet'
```
Comment (rco-odoo) [inline addons/hr_expense/tests/test_expenses.py]: Please add a check that the user cannot read `company_id` on the resulting record.
If this assumption does not hold, the test is useless.

This test should be in base, by the way.

**Comment by tivisse** [comment]
@rco-odoo @sle-odoo 

I've updated the fix + written a test in test_new_api.

If it's ok for you, I would like to proceed to the merge, as this is a blocking issue for the 13.0 migration. 

Thanks.

File: odoo/models.py [hunk 4]
```diff
@@ -3164,11 +3165,11 @@ def _check_company(self, fnames=None):
                 company = self.env.company
             for name in property_fields:
                 # Special case with `res.users` since an user can belong to multiple companies.
```
Comment (Feyensv) [inline odoo/models.py]: corecord missing in this for loop ;).

Comment (tivisse) [reply odoo/models.py]: True. Thanks.

**Comment by tivisse** [comment]
After @sle-odoo approval,

@robodoo r+

Thanks.

**Comment by rco-odoo** [comment]
@tivisse good job!

**Comment by tivisse** [comment]
:boar:

**Pull request closed without merging**